src/boot/efi/xbootldr.c

```
@@ -0,0 +1,211 @@
+                                              where,
+                                              sizeof(gpt_header_buffer), &gpt_header_buffer);
+                        if (EFI_ERROR(r))
+                                continue;
+
+                        /* Some superficial validation of the GPT header */
+                        if(CompareMem(&h->Header.Signature, "EFI PART", sizeof(h->Header.Signature) != 0))
+                                continue;
+
+                        if (h->Header.HeaderSize < 92 ||
+                            h->Header.HeaderSize > 512)
+                                continue;
+
+                        if (h->Header.Revision != 0x00010000U)
+                                continue;
+
+                        /* Calculate CRC check */
+                        crc32_saved = h->Header.CRC32;
+                        h->Header.CRC32 = 0;
+                        r = BS->CalculateCrc32(&gpt_header_buffer, h->Header.HeaderSize, &crc32);
+                        h->Header.CRC32 = crc32_saved;
+                        if (EFI_ERROR(r) || crc32 != crc32_saved)
+                                continue;
+
+                        if (h->MyLBA != where)
+                                continue;
+
+                        if (h->SizeOfPartitionEntry < sizeof(EFI_PARTITION_ENTRY))
+                                continue;
+
+                        if (h->NumberOfPartitionEntries <= 0 ||
+                            h->NumberOfPartitionEntries > 1024)
+                                continue;
+
+                        if (h->SizeOfPartitionEntry > UINTN_MAX / h->NumberOfPartitionEntries) /* overflow check */
+                                continue;
+
+                        /* Now load the GPT entry table */
+                        sz = ALIGN_TO((UINTN) h->SizeOfPartitionEntry * (UINTN) h->NumberOfPartitionEntries, 512);
+                        entries = AllocatePool(sz);
```

we should probably start handling allocation errors sooner or later, btw, i.e. return with `EFI_OUT_OF_RESOURCES` if this failed

---

I am never quite sure when we should just (silently) ignore OOM and try to continue or make it fatal. I had some problem with this in boot.c and deferred that for some other time. Some advice would be nice.

---

hitting oom in inner functions (i.e. "library-like" functions) should just mean the error is propagated as it is, without logging. hitting oom in outer functions (i.e. "main-program-like" functions) should result in logging and also propagation. This is how we typically handle it in the userspace codebase, and I think we should follow the same here too. Thus the errors should typically bubble up all the way to the top, where we should probably catch everything and log loudly and freeze?

---

What if the OOM is caused due to parsing untrusted data (overly large config file). Should that also cause all operation to fail. (This was my primary problem)

---

well, depends... if it's the primary config file then yeah we should bubble the oom case up and then let main() handle it, and freeze. if it's a seconddary config file that is only specific to some small facet, and we can reasonably let the facet fail only then we should log but continue, and skip that facet.

---

Anyway, if you decide to do this, please don't change the logic in the commit that moves the code, but in a later commit.